File: DotNet/CesiumLanguageWriter/Advanced/CesiumFormattingHelper.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Net;
using System.Text;

namespace CesiumLanguageWriter.Advanced
{
    /// <summary>
    /// Contains static methods for formatting data for writing to a CZML stream.
    /// </summary>
    public static class CesiumFormattingHelper
    {
        private static readonly JulianDate s_minimumDate = GregorianDate.MinValue.ToJulianDate();
        private static readonly JulianDate s_maximumDate = GregorianDate.MaxValue.ToJulianDate();

        /// <summary>
        /// Converts a <see cref="TimeInterval"/> as an ISO8601 interval string.
        /// </summary>
        /// <param name="start">The start of the interval.</param>
        /// <param name="stop">The end of the interval.</param>
        /// <param name="format">The format to use.</param>
        /// <returns>The interval represented as an ISO8601 interval string.</returns>
        public static string ToIso8601Interval(JulianDate start, JulianDate stop, Iso8601Format format)
        {
            return ToIso8601(start, format) + "/" + ToIso8601(stop, format);
        }

        /// <summary>
        /// Converts a <see cref="TimeInterval"/> as an ISO8601 interval string.
        /// </summary>
        /// <param name="interval">The interval to convert.</param>
        /// <param name="format">The format to use.</param>
        /// <returns>The interval represented as an ISO8601 interval string.</returns>
        public static string ToIso8601Interval(TimeInterval interval, Iso8601Format format)
        {
            return ToIso8601Interval(interval.Start, interval.Stop, format);
        }

        /// <summary>
        /// Converts a <see cref="JulianDate"/> to an ISO8601 date string.
        /// </summary>
        /// <param name="date">The date to convert.</param>
        /// <param name="format">The format to use.</param>
        /// <returns>The date represented as an ISO8601 date string.</returns>
        public static string ToIso8601(JulianDate date, Iso8601Format format)
        {
            //If the JulianDate is outside the range of supported CZML values,
            //clamp it to the minimum/maximum CZML ISO8601 value.
            if (date <= s_minimumDate)
            {
                return GregorianDate.MinValue.ToIso8601String(format);
            }

            if (date >= s_maximumDate)
            {
                return GregorianDate.MaxValue.ToIso8601String(format);
            }

            return date.ToGregorianDate().ToIso8601String(format);
        }

        /// <summary>
        /// Downloads and converts a remote resource URI into a data URI in the form
        /// <c>data:&lt;MimeType&gt;;base64,&lt;ImageData&gt;</c>, where
        /// <c>&lt;MimeType&gt;</c> is the MIME type of the specified resource, and
        /// <c>&lt;ImageData&gt;</c> is the data encoded as a Base 64 string.
        /// </summary>
        /// <param name="uri">The URI of the resource to convert.</param>
        /// <returns>A data URI containing the content of the resource.</returns>
        public static string DownloadUriIntoDataUri(string uri)
        {
            if (uri.StartsWith("data:"))
                return uri;

            WebRequest request = WebRequest.Create(uri);
            HttpWebRequest httpWebRequest = request as HttpWebRequest;
            if (httpWebRequest != null)
            {
                httpWebRequest.UserAgent = "CesiumWriter";
            }

            using (WebResponse webResponse = request.GetResponse())
            using (Stream responseStream = webResponse.GetResponseStream())
            {
                string mimeType = webResponse.ContentType;
                return BuildDataUri(mimeType, responseStream);
            }
        }

        /// <summary>
        /// Reads from a stream containing an image into a data URI in the form
        /// <c>data:&lt;MimeType&gt;;base64,&lt;ImageData&gt;</c>, where
        /// <c>&lt;MimeType&gt;</c> is the MIME type of the specified image format, and
        /// <c>&lt;ImageData&gt;</c> is the image data encoded as a Base 64 string.
        /// This method does not attempt to decode the image data, but simply writes it directly to the data URI.
        /// </summary>
        /// <param name="stream">The stream containing the image to encode into a data URI.</param>
        /// <param name="imageFormat">The format of the image, which controls the mime type.</param>
        /// <returns>A data URI containing the content of the image.</returns>
        public static string ImageToDataUri(Stream stream, CesiumImageFormat imageFormat)
        {
            string mimeType = GetMimeTypeFromCesiumImageFormat(imageFormat);
            return BuildDataUri(mimeType, stream);
        }

        /// <summary>
        /// Converts an image to a data URI in the form
        /// <c>data:&lt;MimeType&gt;;base64,&lt;ImageData&gt;</c>, where
        /// <c>&lt;MimeType&gt;</c> is the MIME type of the specified <paramref name="image"/>, and
        /// <c>&lt;ImageData&gt;</c> is the image data encoded as a Base 64 string.
        /// </summary>
        /// <param name="image">The image to convert.</param>
        /// <param name="imageFormat">The format of the image, which controls the mime type.</param>
        /// <returns>A data URI containing the content of the image.</returns>
        public static string ImageToDataUri(Image image, CesiumImageFormat imageFormat)
        {
            string mimeType = GetMimeTypeFromCesiumImageFormat(imageFormat);
            using (MemoryStream stream = new MemoryStream())
            {
                image.Save(stream, CesiumImageFormatToImageFormat(imageFormat));
                stream.Position = 0;
                return BuildDataUri(mimeType, stream);
            }
        }

        private static ImageFormat CesiumImageFormatToImageFormat(CesiumImageFormat imageFormat)
        {
            switch (imageFormat)
            {
                case CesiumImageFormat.Jpeg:
                    return ImageFormat.Jpeg;
                case CesiumImageFormat.Png:
                    return ImageFormat.Png;
                case CesiumImageFormat.Bmp:
                    return ImageFormat.Bmp;
                case CesiumImageFormat.Gif:
                    return ImageFormat.Gif;
                default:
                    throw new ArgumentException(CesiumLocalization.ArgumentTypeInvalid, "imageFormat");
            }
        }

        private static string BuildDataUri(string mimeType, Stream dataStream)
        {
            StringBuilder builder = new StringBuilder();
            builder.Append("data:");
            builder.Append(mimeType);
            builder.Append(";base64,");

            using (MemoryStream memoryStream = new MemoryStream())
            {
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = dataStream.Read(buffer, 0, buffer.Length)) > 0)
                    memoryStream.Write(buffer, 0, bytesRead);

                builder.Append(Convert.ToBase64String(memoryStream.GetBuffer(), 0, (int)memoryStream.Length));
            }

            return builder.ToString();
        }

        private static string GetMimeTypeFromCesiumImageFormat(CesiumImageFormat imageFormat)
        {
            switch (imageFormat)
            {
                case CesiumImageFormat.Jpeg:
                    return "image/jpeg";
                case CesiumImageFormat.Png:
                    return "image/png";
                case CesiumImageFormat.Bmp:
                    return "image/bmp";
                case CesiumImageFormat.Gif:
                    return "image/gif";
                default:
                    throw new ArgumentException(CesiumLocalization.ArgumentTypeInvalid, "imageFormat");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumStripeOrientation"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string StripeOrientationToString(CesiumStripeOrientation value)
        {
            switch (value)
            {
                case CesiumStripeOrientation.Horizontal:
                    return "HORIZONTAL";
                case CesiumStripeOrientation.Vertical:
                    return "VERTICAL";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumHorizontalOrigin"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string HorizontalOriginToString(CesiumHorizontalOrigin value)
        {
            switch (value)
            {
                case CesiumHorizontalOrigin.Left:
                    return "LEFT";
                case CesiumHorizontalOrigin.Center:
                    return "CENTER";
                case CesiumHorizontalOrigin.Right:
                    return "RIGHT";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumVerticalOrigin"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string VerticalOriginToString(CesiumVerticalOrigin value)
        {
            switch (value)
            {
                case CesiumVerticalOrigin.Bottom:
                    return "BOTTOM";
                case CesiumVerticalOrigin.Center:
                    return "CENTER";
                case CesiumVerticalOrigin.Top:
                    return "TOP";
                case CesiumVerticalOrigin.Baseline:
                    return "BASELINE";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumHeightReference"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string HeightReferenceToString(CesiumHeightReference value)
        {
            switch (value)
            {
                case CesiumHeightReference.None:
                    return "NONE";
                case CesiumHeightReference.ClampToGround:
                    return "CLAMP_TO_GROUND";
                case CesiumHeightReference.RelativeToGround:
                    return "RELATIVE_TO_GROUND";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumShadowMode"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string ShadowModeToString(CesiumShadowMode value)
        {
            switch (value)
            {
                case CesiumShadowMode.Disabled:
                    return "DISABLED";
                case CesiumShadowMode.Enabled:
                    return "ENABLED";
                case CesiumShadowMode.CastOnly:
                    return "CAST_ONLY";
                case CesiumShadowMode.ReceiveOnly:
                    return "RECEIVE_ONLY";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumInterpolationAlgorithm"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string InterpolationAlgorithmToString(CesiumInterpolationAlgorithm value)
        {
            switch (value)
            {
                case CesiumInterpolationAlgorithm.Linear:
                    return "LINEAR";
                case CesiumInterpolationAlgorithm.Lagrange:
                    return "LAGRANGE";
                case CesiumInterpolationAlgorithm.Hermite:
                    return "HERMITE";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumExtrapolationType"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string ExtrapolationTypeToString(CesiumExtrapolationType value)
        {
            switch (value)
            {
                case CesiumExtrapolationType.None:
                    return "NONE";
                case CesiumExtrapolationType.Hold:
                    return "HOLD";
                case CesiumExtrapolationType.Extrapolate:
                    return "EXTRAPOLATE";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="ClockRange"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string ClockRangeToString(ClockRange value)
        {
            switch (value)
            {
                case ClockRange.Clamped:
                    return "CLAMPED";
                case ClockRange.Unbounded:
                    return "UNBOUNDED";
                case ClockRange.LoopStop:
                    return "LOOP_STOP";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="ClockStep"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string ClockStepToString(ClockStep value)
        {
            switch (value)
            {
                case ClockStep.SystemClock:
                    return "SYSTEM_CLOCK";
                case ClockStep.SystemClockMultiplier:
                    return "SYSTEM_CLOCK_MULTIPLIER";
                case ClockStep.TickDependent:
                    return "TICK_DEPENDENT";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumLabelStyle"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string LabelStyleToString(CesiumLabelStyle value)
        {
            switch (value)
            {
                case CesiumLabelStyle.Fill:
                    return "FILL";
                case CesiumLabelStyle.Outline:
                    return "OUTLINE";
                case CesiumLabelStyle.FillAndOutline:
                    return "FILL_AND_OUTLINE";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumCornerType"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string CornerTypeToString(CesiumCornerType value)
        {
            switch (value)
            {
                case CesiumCornerType.Rounded:
                    return "ROUNDED";
                case CesiumCornerType.Mitered:
                    return "MITERED";
                case CesiumCornerType.Beveled:
                    return "BEVELED";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumColorBlendMode"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string ColorBlendModeToString(CesiumColorBlendMode value)
        {
            switch (value)
            {
                case CesiumColorBlendMode.Highlight:
                    return "HIGHLIGHT";
                case CesiumColorBlendMode.Replace:
                    return "REPLACE";
                case CesiumColorBlendMode.Mix:
                    return "MIX";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Converts a <see cref="CesiumSensorVolumePortionToDisplay"/> to the corresponding string in a CZML stream.
        /// </summary>
        /// <param name="value">The value to convert.</param>
        /// <returns>The string representing the specified value.</returns>
        public static string SensorVolumePortionToDisplayToString(CesiumSensorVolumePortionToDisplay value)
        {
            switch (value)
            {
                case CesiumSensorVolumePortionToDisplay.Complete:
                    return "COMPLETE";
                case CesiumSensorVolumePortionToDisplay.AboveEllipsoidHorizon:
                    return "ABOVE_ELLIPSOID_HORIZON";
                case CesiumSensorVolumePortionToDisplay.BelowEllipsoidHorizon:
                    return "BELOW_ELLIPSOID_HORIZON";
                default:
                    throw new ArgumentException(CesiumLocalization.UnknownEnumerationValue, "value");
            }
        }

        /// <summary>
        /// Returns a resolved url, using the given <see cref="CesiumResourceBehavior"/>.
        /// </summary>
        /// <param name="uri">The url of the resource.</param>
        /// <param name="resourceBehavior">A <see cref="CesiumResourceBehavior"/> specifying how include the resource into a CZML document.</param>
        /// <returns>The resolved url.</returns>
        public static string GetResourceUri(string uri, CesiumResourceBehavior resourceBehavior)
        {
            if (resourceBehavior == CesiumResourceBehavior.Embed)
                return CachingCesiumUriResolver.ThreadLocalInstance.ResolveUri(uri);

            return uri;
        }
    }
}

File: Java/CesiumLanguageWriter/translatedSrc/cesiumlanguagewriter/advanced/CesiumFormattingHelper.java
package cesiumlanguagewriter.advanced;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.ArgumentException;
import agi.foundation.compatibility.ArgumentNullException;
import agi.foundation.compatibility.ConvertHelper;
import agi.foundation.compatibility.HttpWebRequest;
import agi.foundation.compatibility.ImageFormat;
import agi.foundation.compatibility.ImageHelper;
import agi.foundation.compatibility.MemoryStream;
import agi.foundation.compatibility.StreamHelper;
import agi.foundation.compatibility.StringComparison;
import agi.foundation.compatibility.StringHelper;
import agi.foundation.compatibility.Using;
import agi.foundation.compatibility.WebRequest;
import agi.foundation.compatibility.WebResponse;
import cesiumlanguagewriter.*;
import java.awt.image.RenderedImage;
import java.io.InputStream;
import javax.annotation.Nonnull;

/**
 * Contains static methods for formatting data for writing to a CZML stream.
 */
@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
public final class CesiumFormattingHelper {
    private CesiumFormattingHelper() {}

    @Nonnull
    private static JulianDate s_minimumDate = GregorianDate.MinValue.toJulianDate();
    @Nonnull
    private static JulianDate s_maximumDate = GregorianDate.MaxValue.toJulianDate();

    /**
    * Converts a {@link TimeInterval} as an ISO8601 interval string.
    * @param start The start of the interval.
    * @param stop The end of the interval.
    * @param format The format to use.
    * @return The interval represented as an ISO8601 interval string.
    */
    @Nonnull
    public static String toIso8601Interval(@Nonnull JulianDate start, @Nonnull JulianDate stop, @Nonnull Iso8601Format format) {
        return toIso8601(start, format) + "/" + toIso8601(stop, format);
    }

    /**
    * Converts a {@link TimeInterval} as an ISO8601 interval string.
    * @param interval The interval to convert.
    * @param format The format to use.
    * @return The interval represented as an ISO8601 interval string.
    */
    @Nonnull
    public static String toIso8601Interval(@Nonnull TimeInterval interval, @Nonnull Iso8601Format format) {
        if (interval == null) {
            throw new ArgumentNullException("interval");
        }
        return toIso8601Interval(interval.getStart(), interval.getStop(), format);
    }

    /**
    * Converts a {@link JulianDate} to an ISO8601 date string.
    * @param date The date to convert.
    * @param format The format to use.
    * @return The date represented as an ISO8601 date string.
    */
    @Nonnull
    public static String toIso8601(@Nonnull JulianDate date, @Nonnull Iso8601Format format) {
        //If the JulianDate is outside the range of supported CZML values,
        //clamp it to the minimum/maximum CZML ISO8601 value.
        if (JulianDate.lessThanOrEqual(date, s_minimumDate)) {
            return GregorianDate.MinValue.toIso8601String(format);
        }
        if (JulianDate.greaterThanOrEqual(date, s_maximumDate)) {
            return GregorianDate.MaxValue.toIso8601String(format);
        }
        return date.toGregorianDate().toIso8601String(format);
    }

    /**
    * Downloads and converts a remote resource URI into a data URI in the form
    {@code data:<MimeType>;base64,<ImageData>}, where
    {@code <MimeType>} is the MIME type of the specified resource, and
    {@code <ImageData>} is the data encoded as a Base 64 string.
    * @param uri The URI of the resource to convert.
    * @return A data URI containing the content of the resource.
    */
    @Nonnull
    public static String downloadUriIntoDataUri(@Nonnull String uri) {
        if (uri == null) {
            throw new ArgumentNullException("uri");
        }
        if (StringHelper.startsWith(uri, "data:", StringComparison.ORDINAL)) {
            return uri;
        }
        WebRequest webRequest = WebRequest.create(uri);
        HttpWebRequest httpWebRequest = webRequest instanceof HttpWebRequest ? (HttpWebRequest) webRequest : null;
        if (httpWebRequest != null) {
            httpWebRequest.setUserAgent("CesiumWriter");
        }
        try (Using<WebResponse> using$0 = new Using<WebResponse>(webRequest.getResponse())) {
            final WebResponse webResponse = using$0.resource;
            try (Using<InputStream> using$1 = new Using<InputStream>(webResponse.getResponseStream())) {
                final InputStream stream = using$1.resource;
                if (stream == null) {
                    throw new IllegalStateException(CesiumLocalization.getDataDownloadFailed());
                }
                String mimeType = webResponse.getContentType();
                return streamToDataUri(stream, mimeType);
            }
        }
    }

    /**
    * Reads from a stream containing an image into a data URI in the form
    {@code data:<MimeType>;base64,<ImageData>}, where
    {@code <MimeType>} is the MIME type of the specified image format, and
    {@code <ImageData>} is the image data encoded as a Base 64 string.
    This method does not attempt to decode the image data, but simply writes it directly to the data URI.
    * @param stream The stream containing the image to encode into a data URI.
    * @param imageFormat The format of the image, which controls the mime type.
    * @return A data URI containing the content of the image.
    */
    @Nonnull
    public static String imageToDataUri(@Nonnull InputStream stream, @Nonnull CesiumImageFormat imageFormat) {
        String mimeType = getMimeTypeFromCesiumImageFormat(imageFormat);
        return streamToDataUri(stream, mimeType);
    }

    /**
    * Converts an image to a data URI in the form
    {@code data:<MimeType>;base64,<ImageData>}, where
    {@code <MimeType>} is the MIME type of the specified {@code image}, and
    {@code <ImageData>} is the image data encoded as a Base 64 string.
    * @param image The image to convert.
    * @param imageFormat The format of the image, which controls the mime type.
    * @return A data URI containing the content of the image.
    */
    @Nonnull
    public static String imageToDataUri(@Nonnull RenderedImage image, @Nonnull CesiumImageFormat imageFormat) {
        String mimeType = getMimeTypeFromCesiumImageFormat(imageFormat);
        try (Using<MemoryStream> using$0 = new Using<MemoryStream>(new MemoryStream())) {
            final MemoryStream stream = using$0.resource;
            ImageHelper.save(image, stream, cesiumImageFormatToImageFormat(imageFormat));
            stream.setPosition(0L);
            return streamToDataUri(stream, mimeType);
        }
    }

    @Nonnull
    private static ImageFormat cesiumImageFormatToImageFormat(@Nonnull CesiumImageFormat imageFormat) {
        switch (imageFormat) {
        case JPEG: {
            return ImageFormat.getJpeg();
        }
        case PNG: {
            return ImageFormat.getPng();
        }
        case BMP: {
            return ImageFormat.getBmp();
        }
        case GIF: {
            return ImageFormat.getGif();
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getArgumentTypeInvalid(), "imageFormat");
        }
        }
    }

    /**
    * Reads data from a stream into a data URI in the form
    {@code data:<MimeType>;base64,<ImageData>}, where
    {@code <MimeType>} is the given MIME type, and
    {@code <ImageData>} is the image data encoded as a Base 64 string.
    * @param stream The stream to read from.
    * @param mimeType The mime type of the data in the stream.
    * @return A data URI containing the contents of the stream.
    */
    @Nonnull
    public static String streamToDataUri(@Nonnull InputStream stream, @Nonnull String mimeType) {
        if (stream == null) {
            throw new ArgumentNullException("stream");
        }
        if (mimeType == null) {
            throw new ArgumentNullException("mimeType");
        }
        StringBuilder builder = new StringBuilder();
        builder.append("data:");
        builder.append(mimeType);
        builder.append(";base64,");
        try (Using<MemoryStream> using$0 = new Using<MemoryStream>(new MemoryStream())) {
            final MemoryStream memoryStream = using$0.resource;
            StreamHelper.copyTo(stream, memoryStream, 8 * 1024);
            builder.append(ConvertHelper.toBase64String(memoryStream.getBuffer(), 0, (int) memoryStream.getLength()));
        }
        return builder.toString();
    }

    @Nonnull
    private static String getMimeTypeFromCesiumImageFormat(@Nonnull CesiumImageFormat imageFormat) {
        switch (imageFormat) {
        case JPEG: {
            return "image/jpeg";
        }
        case PNG: {
            return "image/png";
        }
        case BMP: {
            return "image/bmp";
        }
        case GIF: {
            return "image/gif";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getArgumentTypeInvalid(), "imageFormat");
        }
        }
    }

    /**
    * Converts a {@link CesiumStripeOrientation} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String stripeOrientationToString(@Nonnull CesiumStripeOrientation value) {
        switch (value) {
        case HORIZONTAL: {
            return "HORIZONTAL";
        }
        case VERTICAL: {
            return "VERTICAL";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumHorizontalOrigin} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String horizontalOriginToString(@Nonnull CesiumHorizontalOrigin value) {
        switch (value) {
        case LEFT: {
            return "LEFT";
        }
        case CENTER: {
            return "CENTER";
        }
        case RIGHT: {
            return "RIGHT";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumVerticalOrigin} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String verticalOriginToString(@Nonnull CesiumVerticalOrigin value) {
        switch (value) {
        case BOTTOM: {
            return "BOTTOM";
        }
        case CENTER: {
            return "CENTER";
        }
        case TOP: {
            return "TOP";
        }
        case BASELINE: {
            return "BASELINE";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumHeightReference} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String heightReferenceToString(@Nonnull CesiumHeightReference value) {
        switch (value) {
        case NONE: {
            return "NONE";
        }
        case CLAMP_TO_GROUND: {
            return "CLAMP_TO_GROUND";
        }
        case RELATIVE_TO_GROUND: {
            return "RELATIVE_TO_GROUND";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumShadowMode} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String shadowModeToString(@Nonnull CesiumShadowMode value) {
        switch (value) {
        case DISABLED: {
            return "DISABLED";
        }
        case ENABLED: {
            return "ENABLED";
        }
        case CAST_ONLY: {
            return "CAST_ONLY";
        }
        case RECEIVE_ONLY: {
            return "RECEIVE_ONLY";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumInterpolationAlgorithm} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String interpolationAlgorithmToString(@Nonnull CesiumInterpolationAlgorithm value) {
        switch (value) {
        case LINEAR: {
            return "LINEAR";
        }
        case LAGRANGE: {
            return "LAGRANGE";
        }
        case HERMITE: {
            return "HERMITE";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumExtrapolationType} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String extrapolationTypeToString(@Nonnull CesiumExtrapolationType value) {
        switch (value) {
        case NONE: {
            return "NONE";
        }
        case HOLD: {
            return "HOLD";
        }
        case EXTRAPOLATE: {
            return "EXTRAPOLATE";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link ClockRange} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String clockRangeToString(@Nonnull ClockRange value) {
        switch (value) {
        case CLAMPED: {
            return "CLAMPED";
        }
        case UNBOUNDED: {
            return "UNBOUNDED";
        }
        case LOOP_STOP: {
            return "LOOP_STOP";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link ClockStep} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String clockStepToString(@Nonnull ClockStep value) {
        switch (value) {
        case SYSTEM_CLOCK: {
            return "SYSTEM_CLOCK";
        }
        case SYSTEM_CLOCK_MULTIPLIER: {
            return "SYSTEM_CLOCK_MULTIPLIER";
        }
        case TICK_DEPENDENT: {
            return "TICK_DEPENDENT";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumLabelStyle} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String labelStyleToString(@Nonnull CesiumLabelStyle value) {
        switch (value) {
        case FILL: {
            return "FILL";
        }
        case OUTLINE: {
            return "OUTLINE";
        }
        case FILL_AND_OUTLINE: {
            return "FILL_AND_OUTLINE";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumArcType} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String arcTypeToString(@Nonnull CesiumArcType value) {
        switch (value) {
        case NONE: {
            return "NONE";
        }
        case GEODESIC: {
            return "GEODESIC";
        }
        case RHUMB: {
            return "RHUMB";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumCornerType} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String cornerTypeToString(@Nonnull CesiumCornerType value) {
        switch (value) {
        case ROUNDED: {
            return "ROUNDED";
        }
        case MITERED: {
            return "MITERED";
        }
        case BEVELED: {
            return "BEVELED";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumClassificationType} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String classificationTypeToString(@Nonnull CesiumClassificationType value) {
        switch (value) {
        case TERRAIN: {
            return "TERRAIN";
        }
        case CESIUM3DTILE: {
            return "CESIUM_3D_TILE";
        }
        case BOTH: {
            return "BOTH";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumColorBlendMode} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String colorBlendModeToString(@Nonnull CesiumColorBlendMode value) {
        switch (value) {
        case HIGHLIGHT: {
            return "HIGHLIGHT";
        }
        case REPLACE: {
            return "REPLACE";
        }
        case MIX: {
            return "MIX";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Converts a {@link CesiumSensorVolumePortionToDisplay} to the corresponding string in a CZML stream.
    * @param value The value to convert.
    * @return The string representing the specified value.
    */
    @Nonnull
    public static String sensorVolumePortionToDisplayToString(@Nonnull CesiumSensorVolumePortionToDisplay value) {
        switch (value) {
        case COMPLETE: {
            return "COMPLETE";
        }
        case ABOVE_ELLIPSOID_HORIZON: {
            return "ABOVE_ELLIPSOID_HORIZON";
        }
        case BELOW_ELLIPSOID_HORIZON: {
            return "BELOW_ELLIPSOID_HORIZON";
        }
        default: {
            throw new ArgumentException(CesiumLocalization.getUnknownEnumerationValue(), "value");
        }
        }
    }

    /**
    * Returns a resolved url, using the given {@link CesiumResourceBehavior}.
    * @param uri The url of the resource.
    * @param resourceBehavior A {@link CesiumResourceBehavior} specifying how include the resource into a CZML document.
    * @return The resolved url.
    */
    @Nonnull
    public static String getResourceUri(@Nonnull String uri, @Nonnull CesiumResourceBehavior resourceBehavior) {
        if (uri == null) {
            throw new ArgumentNullException("uri");
        }
        if (resourceBehavior == CesiumResourceBehavior.EMBED) {
            return CachingCesiumUriResolver.getThreadLocalInstance().resolveUri(uri);
        }
        return uri;
    }
}